### PR TITLE
Implemented SA1133 (incl. tests + codefix)

### LINKED ISSUE #805: Implement SA1133 (DoNotCombineAttributes)
**Category:** Readability
**Title:** Do not combine attributes
**Description:** C# allows attributes to be combined, such as the following:

```csharp
[Export, Shared]
```

This form makes it difficult to distinguish which attributes are applied to a code element, and which arguments are passed to those attributes. To avoid this situation, only place one attribute within a single set of brackets, such as the following:

```csharp
[Export]
[Shared]
```

**Note:** A rule to ensure only one attribute is placed on a given line will be proposed separately.

### PR DESCRIPTION
**Pull request by vweijsters**
Fixes #805

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1133DoNotCombineAttributes.cs [hunk 1]
```diff
@@ -32,7 +38,22 @@ internal class SA1133DoNotCombineAttributes : DiagnosticAnalyzer
...
+            context.RegisterSyntaxNodeActionHonorExclusions(HandleAttributeListAction, SyntaxKind.AttributeList);
+        }
+
+        private static void HandleAttributeList(SyntaxNodeAnalysisContext context)
+        {
+            AttributeListSyntax attributeList = (AttributeListSyntax)context.Node;
+
+            if (attributeList.Attributes.Count > 1)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, attributeList.Attributes[1].GetLocation()));
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1133DoNotCombineAttributes.cs]: :bulb: I believe this highlights the entire argument list for the attribute if one is present. I would prefer to see it only mark the (possibly qualified) attribute type name.

**Comment by sharwell** [comment]
:bulb: Consider including a test which includes multiple attribute lists, each of which contains multiple attributes (bonus for trivia placed between them).

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            var attributeList = (AttributeListSyntax)violatingAttribute.Parent;
+            var newAttributeLists = new List<AttributeListSyntax>();
+
+            var indentationOptions = IndentationOptions.FromDocument(document);
+            var indentationSteps = IndentationHelper.GetIndentationSteps(indentationOptions, attributeList);
+            var indentationTrivia = IndentationHelper.GenerateWhitespaceTrivia(indentationOptions, indentationSteps);
+
+            for (var i = 0; i < attributeList.Attributes.Count; i++)
+            {
+                var newAttributes = SyntaxFactory.SeparatedList(new[] { attributeList.Attributes[i] });
```
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: I'm not sure but `SyntaxFactory.SingletonSeparatedList` could be  used ?

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                var newAttributeList = SyntaxFactory.AttributeList(attributeList.Target, newAttributes);
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: You have to make sure you also set the attribute target.

Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: @pdelvo Is that really necessary? I'm not creating new instances, merely adding the existing ones to new lists.

**Comment by vweijsters** [comment]
Processed review comments

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs (hunk 3 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SA1133CodeFixProvider.cs]: I overlooked that you actually pass the target. So everything is fine

**Comment by sharwell** [comment]
:+1:

**Pull request merged**